jwst/assign_wcs/nircam.py

```diff
@@ -54,14 +70,271 @@ def imaging_distortion(input_model, reference_files):
+    which contains the filter names.
+
+
+    Step 1: Convert the source catalog from the reference frame of the
+            uberimage to that of the dispersed image.  For the Vanilla
+            Pipeline we assume that the pointing information in the file
+            headers is sufficient.  This will be strictly true if all images
+            were obtained in a single visit (same guide stars).
+    Step 2: Record source information for each object in the catalog: position
+            (RA and Dec), shape (A_IMAGE, B_IMAGE, THETA_IMAGE), and all
+            available magnitudes.
+    Step 3: Compute the trace and wavelength solutions for each object in the
+            catalog and for each spectral order.  Record this information.
+    Step 4: Compute the WIDTH of each spectral subwindow, which may be fixed or
+            variable (see discussion of optimal extraction, below).  Record
+            this information.
+    Step 4: Record the MMAG_EXTRACT for each object and spectral order.
+
+    """
+
+    # The input is the grism image
+    if not isinstance(input_model, ImageModel):
+        raise TypeError('The input data model must be an ImageModel.')
+
+    # make sure this is a grism image
+    if "NRC_GRISM" not in input_model.meta.exposure.type:
+            raise TypeError('The input exposure is not a NIRCAM grism')
+
+    # Create the empty detector as a 2D coordinate frame in pixel units
+    gdetector = cf.Frame2D(name='grism_detector',
+                           axes_order=(0, 1),
+                           unit=(u.pix, u.pix))
+
+    # translate the x,y detector-in to x,y detector out coordinates
+    # Get the disperser parameters which are defined as a model for each
+    # spectral order
+    log.info("Collecting specwcs information")
+    with NIRCAMGrismModel(reference_files['specwcs']) as f:
+        xcoeff = f.xcoeff
+        ycoeff = f.ycoeff
+        lcoeff = f.lcoeff
+        orders = [j for j, k in enumerate(f.orders)]  # ordered [str names]
+        wrange_selector = f.wrange_selector
+        wrange = f.wrange
+
+    # add the wavelength range using the far left and far right orders?
+    fselect1 = wrange_selector[0].index(input_model.meta.instrument.filter)
+    fselect2 = wrange_selector[-1].index(input_model.meta.instrument.filter)
+    lower_lam = wrange[0][fselect1][0]
+    upper_lam = wrange[-1][fselect2][1]
+    input_model.meta.wcsinfo.waverange_start = lower_lam
+    input_model.meta.wcsinfo.waverange_end = upper_lam
+
+    # could add wrange input here to track user wavelength input validity
+    nbtrace = NIRCAMGrismModelSelector(orders,
+                                       xcoeff=xcoeff,
+                                       ycoeff=ycoeff,
+                                       lcoeff=lcoeff,
+                                       name='grism2image')
+
+    # (x,y,l,spec_order)->through grism trace polynomials -> x, y
+    det2det = (Identity(4)) | nbtrace
```

Comment:
Possibly, I noticed a bug in my model logic this afternoon, this may change when I figure out how to deal with it. Lemme ping you again with the next push?